src/Compilers/CSharp/Portable/Binder/Semantics/BestTypeInferrer.cs

```diff
@@ -96,7 +88,7 @@ public static TypeSymbol InferBestType(ImmutableArray<BoundExpression> exprs, Co
             // Perform best type inference on candidate types.
             var builder = ArrayBuilder<TypeSymbol>.GetInstance(candidateTypes.Count);
             builder.AddRange(candidateTypes);
-            var result = GetBestType(builder, conversions, ref useSiteDiagnostics);
+            var result = GetBestType(builder, conversions, out hadNullabilityMismatch, ref useSiteDiagnostics);
```

Comment:
>out hadNullabilityMismatch [](start = 59, length = 26)

Something seems wrong with this approach. If it is used during initial binding, it should never return hadNullabilityMismatch, or at least it is meaningless to the caller. But this method can't be used after initial binding, as it uses `expr.Type` for the type rather than the type that was computed by flow analysis for the nullable-aware type of the constituent expressions. So based on my reading of the method, something is wrong for either potential caller.

Reply:
Added an assert to `GetBestType` that `!hadNullabilityMismatch || conversions.IncludeNullability`. And `NullableWalker.VisitArrayInitializer` will generate `BoundExpressionWithNullability` expressions that contain inferred nested nullability.

---
In reply to: [225349914](https://github.com/dotnet/roslyn/pull/29727#discussion_r225349914) [](ancestors = 225349914)